src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -122,58 +125,99 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
-
-            const allStudentEntries: StudentEnrollRequest[] = [];
-            allStudentChunks.forEach((chunk: StudentEnrollRequest[]) => allStudentEntries.push(...chunk));
-
-            this.enrollResultPanelList =
-                this.populateEnrollResultPanelList(this.existingStudents, enrolledStudents, allStudentEntries);
-
-            this.studentService.getStudentsFromCourse({ courseId: this.courseId }).subscribe((resp: Students) => {
-              this.existingStudents = resp.students;
-              if (!this.isExistingStudentsPanelCollapsed) {
-                const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
-                this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
-              }
-              this.isExistingStudentsPresent = true;
-            });
-          },
-          error: (resp: ErrorMessageOutput) => {
-            this.enrollErrorMessage = resp.error.message;
-          },
+    );
+
+    const handledRequests: Observable<Students> = this.handleTimeoutForEnrollRequest(enrollRequests);
+    handledRequests.subscribe({
+      next: (resp: Students) => {
+        enrolledStudents.push(...resp.students);
+        this.remainingStudentChunks = this.remainingStudentChunks.slice(1, this.remainingStudentChunks.length);
+      },
+      complete: () => {
+        this.showEnrollResults = true;
+        this.statusMessage.pop(); // removes any existing error status message
+        this.statusMessageService.showSuccessToast('Enrollment successful. Summary given below.');
+
+        const allStudentEntries: StudentEnrollRequest[] = [];
+        this.allStudentChunks.forEach((chunk: StudentEnrollRequest[]) => allStudentEntries.push(...chunk));
+
+        this.enrollResultPanelList =
+            this.populateEnrollResultPanelList(this.existingStudents, enrolledStudents, allStudentEntries);
+
+        this.studentService.getStudentsFromCourse({ courseId: this.courseId }).subscribe((resp: Students) => {
+          this.existingStudents = resp.students;
+          if (!this.isExistingStudentsPanelCollapsed) {
+            const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
+            this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
+          }
+          this.isExistingStudentsPresent = true;
         });
+      },
+      error: (resp: ErrorMessageOutput) => {
+        this.enrollErrorMessage = resp.error.message;
+      },
+    });
   }
 
+  private handleTimeoutForEnrollRequest(prevRequests: Observable<Students>): Observable<Students> {
+
+    return prevRequests.pipe(finalize(() => this.isEnrolling = false))
+        .pipe(takeWhile(() => this.remainingStudentChunks.length > 0))
+        .pipe(catchError(() => {
+          // First request in the list caused the timeout
+          const currentLongRequest: StudentEnrollRequest[] = this.remainingStudentChunks[0];
```

Comment:
it will retry sending if it fails regardless of the error. I will add a check to limit to at most 1 retry

Reply:
Oh what I mean is that upon catching the error, can the logic here be sure that the error is BE timeout error? Or some other form of error e.g. connection timeout error (from TCP socket), unauthorized error (idk how it could be, just an example)?

Limiting the # of retries is good practice. Abort when a certain call exceed the cap. Can set it higher honestly due to the idempotent enrolment in the BE.

Reply:
I can check the response error message to see if its timeout or not. But I'm not sure what the message is because its returned by GAE.

I think limiting the number of retries should suffice, since we would like to retry sending upon failure of any kind

Reply:
I get your logic. I don't really have a problem should there be a guarantee that there won't be an edge case of novel backend error that would cause the enrolment workflow to go into invalid state.